src/dbnode/storage/index/fields_terms_iterator.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"github.com/m3db/m3/src/m3ninx/index/segment"
+	xerrors "github.com/m3db/m3/src/x/errors"
+)
+
+// fieldsAndTermsIterator iterates over all known fields and terms for a segment.
+type fieldsAndTermsIterator interface {
+	// Next returns a bool indicating if there are any more elements.
+	Next() bool
+
+	// Current returns the current element.
+	// NB: the element returned is only valid until the subsequent call to Next().
+	Current() (field, term []byte)
+
+	// Err returns any errors encountered during iteration.
+	Err() error
+
+	// Close releases any resources held by the iterator.
```

Comment:
Maybe add a comment explicitly saying this will not return the iter to the pool and that anything implementing this interface should explicitly support double closes

Reply:
as mentioned earlier, reworked to avoid this dependency.

Reply:
done